packages/flutter_riverpod/test/change_notifier_provider_test.dart

```diff
@@ -78,6 +80,59 @@ void main() {
 
     expect(notifier.mounted, isFalse);
   });
+
+  test('can set state', () async {
+    final container = ProviderContainer();
+    final initialNotifier = TestNotifier();
+    final notifier = TestNotifier();
+    final completer = Completer<void>();
+    final provider = ChangeNotifierProvider((ref) {
+      Future.microtask(() {}).then((value) {
+        ref.setState(notifier);
+        completer.complete();
+      });
+      return initialNotifier;
+    });
+
+    container.read(provider);
+    await completer.future;
+    expect(container.read(provider), notifier);
+  });
+
+  test('dispose', () async {
+    final container = ProviderContainer();
+    final notifier = TestNotifier();
+    final provider = ChangeNotifierProvider((ref) {
+      return notifier;
+    });
+
+    container.read(provider);
+    container.dispose();
+    expect(notifier.mounted, isFalse);
+  });
+
+  test('dispose when used with set state', () async {
+    final container = ProviderContainer();
+    final initialNotifier = TestNotifier();
+    final notifier = TestNotifier();
+    final completer = Completer<void>();
+    final provider = ChangeNotifierProvider((ref) {
+      Future.microtask(() {}).then((value) {
+        ref.setState(notifier);
+        completer.complete();
+      });
+      return initialNotifier;
+    });
+
+    container.read(provider);
+    await completer.future;
+    container.read(provider);
+    expect(initialNotifier.mounted, isFalse);
+    expect(notifier.mounted, isTrue);
+
+    container.dispose();
+    expect(notifier.mounted, isFalse);
```

Comment:
We need to check that the new notifier is listened too (and the previous one no-longer is)